File: dotnet/2015/AdventOfCode.Y2015/Day01.cs
using AdventOfCode.Lib;

namespace AdventOfCode.Y2015;

[AocPuzzle(2015, 1, "Not Quite Lisp")]
public sealed class Day01 : IAocDay<int>
{
    public static int Part1(AocInput input) => input.Text.Sum(GoUpOrDownOneFloor);

    public static int Part2(AocInput input) => input.Text
        .Select(GoUpOrDownOneFloor)
        .Aggregate(new List<int>(), (floors, floor) =>
        {
            var lastFloorLevel = floors.Count > 0 ? floors[^1] : 0;
            var updatedFloor = lastFloorLevel + floor;
            floors.Add(updatedFloor);
            return floors;
        })
        .IndexOf(-1) + 1;

    private static int GoUpOrDownOneFloor(char x) => x is '(' ? 1 : -1;
}

File: python/2015/day01.py
from aoc.core import solve, AocInput


def goUpOrDownOneFloor(c: str) -> int:
    return 1 if c == '(' else - 1


def part1(input: AocInput) -> int:
    return sum(list(map(goUpOrDownOneFloor, input.text)))


def part2(input: AocInput) -> int:
    floors = []
    res = list(map(goUpOrDownOneFloor, input.text))
    for floor in res:
        lastFloorLevel = floors[-1] if len(floors) > 0 else 0
        updatedFloor = lastFloorLevel + floor
        floors.append(updatedFloor)

    return floors.index(-1) + 1


def main() -> None:
    DAY = 'Day01'
    solve(DAY, part1)
    solve(DAY, part2)


if __name__ == '__main__':
    main()
